File: haskell/perfect-numbers/src/PerfectNumbers.hs
module PerfectNumbers (classify, Classification(..)) where

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)

classify :: Int -> Maybe Classification
classify num
  | num < 1 = Nothing
classify num 
  | s == num = Just Perfect
  | s < num = Just Deficient
  | otherwise = Just Abundant
  where s = foldr (\x acc -> if num `mod` x == 0 then acc + x else acc) 0 [1..(div num 2)]

File: python/perfect-numbers/perfect_numbers.py
from functools import reduce
def classify(number):
    if number <= 0:
        raise ValueError("Non positive number")
    aliquot = reduce(lambda s, d: s + d if number % d == 0 else s, range(1, number//2 + 1), 0)
    if aliquot == number:
        return 'perfect'
    elif aliquot < number:
        return 'deficient'
    else:
        return 'abundant'
